File: Semestrul 3/Metode Avansate de Programare/Curs/curs3/src/main/java/ro/ubbcluj/map/repository/InMemoryRepository.java
package ro.ubbcluj.map.repository;

import ro.ubbcluj.map.domain.Entity;
import ro.ubbcluj.map.domain.validators.Validator;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID,E> {
    private Validator<E> validator;
    Map<ID,E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities=new HashMap<ID,E>();
    }

    @Override
    public E findOne(ID id){
        if (id==null)
            throw new IllegalArgumentException("id must be not null");
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if (entity==null)
            throw new IllegalArgumentException("entity must be not null");
        validator.validate(entity);
        if(entities.get(entity.getId()) != null) {
            return entity;
        }
        else entities.put(entity.getId(),entity);
        return null;
    }

    @Override
    public E delete(ID id) {
        return null;
    }

    @Override
    public E update(E entity) {

        if(entity == null)
            throw new IllegalArgumentException("entity must be not null!");
        validator.validate(entity);

        entities.put(entity.getId(),entity);

        if(entities.get(entity.getId()) != null) {
            entities.put(entity.getId(),entity);
            return null;
        }
        return entity;

    }
}

File: Semestrul 3/Metode Avansate de Programare/Seminar/sem11_ex/repository/InMemoryRepository.cs
using sem11.domain;
using sem11.domain.validators;

namespace sem11.repository;


class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
{
    protected IValidator<E> vali;

    protected IDictionary<ID, E> entities = new Dictionary<ID, E>();

    public InMemoryRepository(IValidator<E> vali)
    {
        this.vali = vali;
    }

    public E Delete(ID id)
    {
        throw new NotImplementedException();
    }

    public IEnumerable<E> FindAll()
    {
        return entities.Values.ToList<E>();
    }

    public E FindOne(ID id)
    {
        throw new NotImplementedException();
    }

    public E Save(E entity)
    {
        if (entity == null)
            throw new ArgumentNullException("entity must not be null");
        this.vali.Validate(entity);
        if (this.entities.ContainsKey(entity.Id))
        {
            return entity;
        }

        this.entities[entity.Id] = entity;
        return default(E);
    }

    public E Update(E entity)
    {
        throw new NotImplementedException();
    }
}

